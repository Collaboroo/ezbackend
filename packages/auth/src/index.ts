import 'fastify-passport';
export * from './auth';
export * from './ez-user';
export * from './providers';
