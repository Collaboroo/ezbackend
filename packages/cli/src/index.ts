import { Command } from "commander";
import { sync as readPackageUpSync } from "read-pkg-up";
import initiate from "./initiate";

//URGENT TODO: Make dependencies core dependencies not dev dependencies during installation
//URGENT TODO: Add @ezbackend/core to dependencies

const pkg = readPackageUpSync({ cwd: __dirname })?.packageJson;

const program = new Command();

program
  .command("init")
  .description("Initialize ezbackend into your project.")
  .option("-f --force", "Force add ezbackend")
  .action((options) => initiate(options, pkg));

program.parse(process.argv);

