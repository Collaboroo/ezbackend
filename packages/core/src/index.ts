type IEzbPlugin = () => void;

export type IEzbPlugins = {
  preInit: Array<IEzbPlugin>;
  init: IEzbPlugin | null;
  postInit: Array<IEzbPlugin>;
  preHandler: Array<IEzbPlugin>;
  handler: IEzbPlugin | null;
  postHandler: Array<IEzbPlugin>;
  preRun: Array<IEzbPlugin>;
  run: IEzbPlugin | null;
  postRun: Array<IEzbPlugin>;
};

export class EzBackend {
  plugins: IEzbPlugins;

  private static instance: EzBackend;

  constructor() {
    this.plugins = {
      preInit: [],
      init: () => {},
      postInit: [],
      preHandler: [],
      handler: () => {},
      postHandler: [],
      preRun: [],
      run: () => {},
      postRun: [],
    };
  }

  public static initializeApp() {
    if (!EzBackend.instance) {
      EzBackend.instance = new EzBackend();
    }
  }

  public static app(): EzBackend {
    EzBackend.initializeApp();
    return EzBackend.instance;
  }

  public static start(): void {
    const ezb = EzBackend.app();
    const plugins = ezb.plugins;

    plugins.preInit.forEach((plugin) => {
      plugin();
    });

    plugins.init();

    plugins.postInit.forEach((plugin) => {
      plugin();
    });

    plugins.preHandler.forEach((plugin) => {
      plugin();
    });

    plugins.handler();

    plugins.postHandler.forEach((plugin) => {
      plugin();
    });

    plugins.preRun.forEach((plugin) => {
      plugin();
    });

    plugins.run();

    plugins.postRun.forEach((plugin) => {
      plugin();
    });
  }
}
