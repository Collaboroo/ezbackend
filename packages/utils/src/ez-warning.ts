import chalk from 'chalk';

export function ezWarning(msg: string, logger?: any) {
  if (process.env.NODE_ENV === 'test') return
  const warningMsg = '⚠️ ' + chalk.yellow(msg);
  if (logger) {
    logger.log(warningMsg);
  } else {
    console.log(warningMsg);
  }
}
