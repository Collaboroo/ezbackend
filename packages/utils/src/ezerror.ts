import chalk from "chalk";

function lineFactory(prefix: string, msg: string | undefined) {
  if (msg === undefined) {
    return "";
  } else
    return (
      chalk.bgRed(" " + prefix.padEnd(5, " ")) + chalk.gray(": ") + msg + "\n"
    );
}

export class EzError extends Error {
  summary: string;
  description: string;
  code?: string;
  link?: string;

  constructor(msg: string, desc: string, oneLiner?: string, docs?: string) {
    const errMsg =
      "\n\n" +
      lineFactory("Err", msg) +
      lineFactory("Desc", desc) +
      lineFactory("Fix", oneLiner) +
      lineFactory("Docs", docs);
    super(errMsg);

    this.summary = msg;
    this.description = desc;
    this.code = oneLiner;
    this.link = docs;
  }
}
