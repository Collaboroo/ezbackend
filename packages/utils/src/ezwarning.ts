import chalk from 'chalk'

export function ezWarning(msg:string) {
    console.log('⚠️ ',chalk.yellow(msg))
}