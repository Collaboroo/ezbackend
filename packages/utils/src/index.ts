export * from './ez-error';
export * from './ez-warning';
export * from './overload-parameters-type';
