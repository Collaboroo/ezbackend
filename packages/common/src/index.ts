import { IOptions } from "./definitions";
import { IEzbConfig } from "@ezbackend/core";
import { fastify } from "fastify";
import path from "path";
import { EzBackend } from "./definitions";
import { mixedInstance } from "avvio";
import { createConnection } from "typeorm";
import { RouteOptions } from "fastify";
import { APIGenerator } from "./models";
import {convert} from './models/typeorm-json-schema'

const ezb = EzBackend.app();

//Configure defaults
ezb.plugins.init = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig & IOptions,
  cb
) => {
  ezb.orm = await createConnection(opts.orm);
  ezb.server = fastify(opts.server);
  cb();
};

ezb.plugins.handler = (ezb: mixedInstance<EzBackend>, opts: IEzbConfig, cb) => {
  const customEzbPath =
    opts.entryPoint ?? path.join(process.cwd(), ".ezb/index.ts");
  require(customEzbPath);
  ezb.models.forEach((model) => {
    const repo = ezb.orm.getRepository(model);

    //LEFT OFF
    // const metaData = ezb.orm.getMetadata(model)
    // console.log(convert(metaData))

    const generator = new APIGenerator(repo, { prefix: repo.metadata.name });
    generator.generateRoutes();
  });
  cb();
};

ezb.plugins.run = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig,
  cb
) => {
  await ezb.server.listen(opts.port, function (err, address) {
    if (err) {
      console.error(err);
      process.exit(1);
    }
  });
  cb();
};

export { EzBackend } from "./definitions";
export { EzModel, response, APIGenerator } from "./models";
