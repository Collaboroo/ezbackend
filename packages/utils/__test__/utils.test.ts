import { EzError } from "../src";

describe("EzError", () => {
  describe("Full Error", () => {
    let err: EzError;

    beforeAll(() => {
      err = new EzError(
        "Error Summary",
        "Error Description",
        "Some Code",
        "Docs Link",
      );
    });
    it("Should throw a well formatted error", async () => {
      expect(err.toString()).toContain("Error Summary");
      expect(err.toString()).toContain("Error Description");
      expect(err.toString()).toContain("Some Code");
      expect(err.toString()).toContain("Docs Link");
    });

    it("Should contain a summary property", async () => {
      expect(err.summary).not.toBe(undefined);
      expect(err.summary).toBe("Error Summary");
    });

    it("Should contain a description property", async () => {
      expect(err.description).not.toBe(undefined);
      expect(err.description).toBe("Error Description");
    });

    it("Should contain a code property", async () => {
      expect(err.code).not.toBe(undefined);
      expect(err.code).toBe("Some Code");
    });

    it("Should contain a link property", async () => {
      expect(err.link).not.toBe(undefined);
      expect(err.link).toBe("Docs Link");
    });
  });

  describe("Partial Error", () => {
    let err: EzError;

    beforeAll(() => {
      err = new EzError("Error Summary", "Error Description");
    });
    it("Should work fine without the code", async () => {
      expect(err.code).toBe(undefined);
    });

    it("Should work fine without the link", async () => {
      expect(err.link).toBe(undefined);
    });
  });
});
