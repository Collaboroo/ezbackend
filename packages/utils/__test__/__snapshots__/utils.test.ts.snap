// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EzError Full Error Should throw a well formatted error 1`] = `
"Error: 

[41m Err  [49m[90m: [39mError Summary
[41m Desc [49m[90m: [39mError Description
[41m Fix  [49m[90m: [39mSome Code
[41m Docs [49m[90m: [39mDocs Link
"
`;
